Fix diff backbone issue

devmanhinton opened the pull request:
@yaymukund cc @tobowers 

This is a fix for when the Backbone EndDash uses is different then the Backbone that the client's app uses.

Ie., EndDash's Backbone === Client's Backbone // returns false

I ran into this with my node application which was a hella of a pain to track down but was rewarding once I did!

An important note is, though I do not think this will ever come up, adding a backbone model to a collection from a different backbone will not work properly (this is spec-ed). 

===== Below is fun random fact ==== no need to read ==== repeat fun & AWESOME but not necessary
(More details on last sentence). Backbone collections do ~ if not Backbone.Model || Collection convert it to use Backbone style events (so collections can listen for changes). That means take all top level attributes -> attribute object and do ModelPassedIn.attributes = attributes. With a BackboneModel from a different Backbone this overwrites its attributes with a nested version of itself. In conclusion. Collection.get(model).attributes.attributes is actually the model's attributes. Fun times and likely unncessary.

Commit:
A change to allow for any object that provides a get, set, on, and once function to be used as an EndDash model. Also, a spec that tests it

Commit:
Let the specs continue

Commit:
Collections with mixed model type specs added

yaymukund commented on test/non_backbone_integration.js:
```diff
@@ -0,0 +1,184 @@
+  beforeEach(function(){
+    var backModel = new Backbone.Model()
+  })
+  it("Should work as normal", function(){
```
This test description should be more specific

Commit:
Test messages made clearer

yaymukund commented on lib/reactions/model.js:
```diff
@@ -18,6 +18,14 @@ var ModelReaction = Reaction.extend({
     next(reload)
   },
 
+  isEndDashCompatabile: function(model) {
```
should be `Compatible`

yaymukund commented:
As I said before, we should be using semicolons consistently in new code.

yaymukund commented on test/non_backbone_integration.js:
```diff
@@ -0,0 +1,184 @@
+describe("With a backbone collection from a different backbone object", function(){
+  beforeEach(function(){
+    this.collection = new BackboneClone.Collection()
+    this.templateString = "<div class='users-'><div class ='user-'><div class=' #{hook} persisted-'></div></div></div>"
```
Split up this string over multiple lines so it's easier to read

yaymukund commented on lib/reactions/model.js:
```diff
@@ -18,6 +18,14 @@ var ModelReaction = Reaction.extend({
   },
 
+  isEndDashCompatabile: function(model) {
+    return (typeof model.on == 'function' &&
```
Still not comfortable with this implementation but I think it's good enough for now. I'll open up an issue later with some thoughts on how to do this correctly.

Commit:
Semi-colon-ized and template string split to multiple lines for greater readability

yaymukund commented on test/non_backbone_integration.js:
```diff
@@ -0,0 +1,184 @@
+describe("With a backbone model from a different backbone object", function(){
+  it("binding it to a template should work normally", function(){
+    var model = new BackboneClone.Model({persisted: "Chelsa Piers"})
+      , template = generateTemplate({user: model}, "<div class='user-'><div class='persisted-'></div></div>")
```
Please split this string over multiple lines so it's easier to read. Same goes for all the `generateTemplate` calls in this PR.

yaymukund commented on test/non_backbone_integration.js:
```diff
@@ -0,0 +1,189 @@
+                                        // See lib/reactions/model.js for fix.
+
+describe("With a backbone model from a different backbone object", function(){
+  it("binding it to a template should work normally", function(){
```
These tests are okay, but you can make them more semantic. For example, here's how I'd rewrite this test:

``` javascript
var user = new Backbone.Model({name: 'Chelsa Piers'}),
    template = generateTemplate({user: user}, markup);

expect($('.name').text()).to.be('Chelsa Piers');
user.set('name', 'Larry Gonick');
expect($('.name').text()).to.be('Larry Gonick');
```

Realize that people sometimes skim tests in order to familiarize themselves with the library, so it's doubly important to make them easy to read.

yaymukund commented on test/non_backbone_integration.js:
```diff
@@ -0,0 +1,189 @@
+})
+
+describe("With a backbone model", function(){
+  beforeEach(function(){
```
`backModel` never seems to be used.

yaymukund commented on test/non_backbone_integration.js:
```diff
@@ -0,0 +1,189 @@
+  })
+})
+
+describe("With a backbone collection from a different backbone object", function(){
```
So much copy-pasted code here. Can you please dry it up using `beforeEach` or something?

yaymukund commented on lib/reactions/model.js:
```diff
@@ -2,34 +2,42 @@ var Reaction = require("../reaction")
+            typeof model.once == 'function' &&
+            typeof model.set == 'function' &&
+            typeof model.get == 'function'
+           )
```
`;`

:page_facing_up:

yaymukund commented on lib/reactions/model.js:
```diff
@@ -2,34 +2,42 @@ var Reaction = require("../reaction")
+    next(reload);
+  },
+
+  isEndDashCompatbile: function(model) {
```
Another typo?

Commit:
Classic misspelling, well as long as you are consistent it works. But change it anyways

yaymukund commented:
@devmanhinton Let me know when this stuff is fixed and you want me to review again.

Commit:
Account for a model being undefined, which happens in brief. This should and will be spec-ed

Commit:
Spec now resuses same object and same template strings

Commit:
Semi-colons added to all the lines and use jquery nth-child instead of the hook I was using because this be clearer.

devmanhinton commented:
@yaymukund 

In terms of the spec:

In addition, I moved all the models & templates into a beforeEach top level describe block as well as use JQuery selectors instead of the hook I was using before. Finally, I added a lotta semi-colons. I think this really helps the readability.

In terms of a new bug I found and fixed:

Bug:
Template generated that scopes down to a model or property that is undefined*
lib/reactions/model.js-- calls isCompatible(model)
inside of isCompatible-- calls model.on
--> Exception, cannot call property 'on' of undefined.

Fix:
Add if(!model) return false; to lib/reactions/model.js:isCompatible (commented in this PR)

*note: I'm not sure of the exact cause but is from brief script edit view (creating a new script)

devmanhinton commented on lib/reactions/model.js:
```diff
@@ -2,34 +2,45 @@ var Reaction = require("../reaction")
+  isEndDashCompatible: function(model) {
+    if (!model) {
+      return false;
+    }
```
Above is check if model passed in is undefined

devmanhinton commented:
Missed committing a few semi colons in lib/reaction/model.js. Let me do that after our meeting before merging.

Commit:
Fix up more semi-colons

bglusman commented:
@yaymukund this seems good to merge now, but I'll let you pull trigger since you were reviewer.

yaymukund commented:
The tests could use some cleanup, but this has been sitting around long enough so I'm going to merge.
